auth/src/main/java/org/aerogear/mobile/security/AsyncSecurityCheckExecutor.java

```diff
@@ -94,19 +98,40 @@ public AsyncSecurityCheckExecutor build() {
      * The key of the map will be the output of {@link SecurityCheck#getName()}, while the value will be
      * a {@link Map} of {@link Future} with the {@link SecurityCheckResult} of the check.
      *
+     * @param securityCheckExecutorListeners list of listeners that will receive events about checks execution
      * @return {@link Map}
      */
-    public Map<String, Future<SecurityCheckResult>> execute() {
+    public Map<String, Future<SecurityCheckResult>> execute(SecurityCheckExecutorListener... securityCheckExecutorListeners) {
+
+        final List<SecurityCheckExecutorListener> listeners;
+        final Collection<SecurityCheck> checks = getChecks();
+
+        // Initializes listener list: the metric publisher will be added to passed in listeners
+        if (securityCheckExecutorListeners == null) {
+            listeners = new ArrayList<>(1);
+        } else {
+            listeners = new ArrayList(Arrays.asList(securityCheckExecutorListeners));
+        }
+
+        listeners.add(getMetricServicePublisher());
 
-        final MetricsService metricsService = getMetricsService();
         final Map<String, Future<SecurityCheckResult>> res = new HashMap<>();
+        AtomicInteger count = new AtomicInteger(checks.size());
 
         for (final SecurityCheck check : getChecks()) {
             res.put(check.getName(), (executorService.submit(() -> {
                 final SecurityCheckResult result =  check.test(getContext());
-                if (metricsService != null) {
-                    metricsService.publish(new SecurityCheckResultMetric(result));
+
+                for (SecurityCheckExecutorListener listener : listeners) {
+                    listener.onExecuted(result);
+                }
+
+                if (count.decrementAndGet() == 0) {
+                    for (SecurityCheckExecutorListener listener : listeners) {
+                        listener.onFinished();
+                    }
                 }
+
```

Comment:
this definitely is a clean way to supporting the batch metrics. However, for now, I don't think we want to expose this method to developers as it is making the API quite complicated. This is already beyond the original requirement.

So can I suggest let's make this method private, and create a new public `execute` method that takes no parameters. In the implementation of that method, it will add a metrics listener and call the private method.

This way if this feature proves useful in the future, we can make it public at that point.

WDYT?

Reply:
We already had a similar chat about this ;) I don't think it really adds complexity since Listeners are pretty standard in Java programming, however it won't be a big deal to hide it, so I can certainly do that.